### ISSUE TITLE
LightGBM fails linear classification task

### ISSUE THREAD

**Issue by MatthiasMichael** [microsoft/LightGBM#3401]
I am not filing this as a bug because I am pretty sure the error is on my side.
I am trying to use LightGBM for a small toy classification example (sklearn iris dataset reduced to two classes and two features) that is linearly separable. LightGBM however fails to find an optimal separation between the examples used for training while the GradientBoostingClassifier integrated in sklearn does not seem to have a problem with this task:
![lgbm_sklearn_comparison](https://user-images.githubusercontent.com/47566356/93600062-6104dd80-f9bf-11ea-9bd3-3ad62e729b6c.png)
The only parameter that seems to have a visible effect on the result is the number of trees that should be learned. I am however unable to find any combination of num_trees, num_leaves, max_depth, and learning_rate that allows a perfect fit to the training data. Most other available parameters LightGBM provides are only useful for dealing with overfitting while I am (more or less) underfitting so they do not seem useful to me.

What am I doing wrong?
How can I convince LightGBM to fit to the presented training data?

Thank you in advance.
The behaviour is identical with versions 2.3.1 and 3.0.0.

Here is the code used to generate the above image (the functions are only used to draw the dataset and the decision boundary and can be considered 'noise'):

    from sklearn.datasets import load_iris
    from sklearn.ensemble import GradientBoostingClassifier
    from sklearn.metrics import accuracy_score
    from sklearn.preprocessing import LabelEncoder
    
    from lightgbm.sklearn import LGBMClassifier
    
    from matplotlib.colors import ListedColormap
    
    import matplotlib.pyplot as plt
    import numpy as np
    
    # Functions for drawing classification result and decision boundary
    def meshgrid(X):
        step_size = 0.02
        x_min, x_max = X[:, 0].min() - .5, X[:, 0].max() + .5
        y_min, y_max = X[:, 1].min() - .5, X[:, 1].max() + .5
        return np.meshgrid(np.arange(x_min, x_max, step_size),
                           np.arange(y_min, y_max, step_size))
    
    
    def meshdata(X):
        xx, yy = meshgrid(X)
        return np.c_[xx.ravel(), yy.ravel()]
    
    
    def plot_decision_boundary(ax, cls, X):
        xx, yy = meshgrid(X)
        data = meshdata(X)
    
        Z = cls.predict_proba(data)
        if Z.shape[1] > 1:
            Z = Z[:, 1]
    
        Z = Z.reshape(xx.shape)
        ax.contourf(xx, yy, Z, cmap=plt.cm.RdBu, alpha=0.6)
    
    
    def plot_classification(ax, cls, X, y):
        plot_decision_boundary(ax, cls, X)
    
        y_pred = cls.predict(X)
        score = accuracy_score(y, y_pred)
    
        X_correct = X[y_pred == y]
        X_wrong = X[y_pred != y]
    
        y_correct = y[y_pred == y]
        y_wrong = y[y_pred != y]
    
        ax.scatter(X_correct[:, 0], X_correct[:, 1], c=y_correct, cmap=plt.cm.RdBu, alpha=0.4)
        ax.scatter(X_wrong[:, 0], X_wrong[:, 1], c=y_wrong, s=90, edgecolor='k',
                   cmap=ListedColormap(['#FF0000', '#0000FF']), marker='X')
    
        ax.text(X[:, 0].max() - .3, X[:, 1].min() - .2, ('%0.2f' % score).lstrip('0'),
                size=15, horizontalalignment='right')
    
    
    # Training the classifiers
    n_estimators = 1
    n_leaves = 8
    max_depth = 4
    learning_rate = 1
    
    classes = [1, 0]
    features = [2, 0]
    
    iris = load_iris()
    
    mask = [x in classes for x in iris.target]
    
    X = iris.data[mask][:, features]
    y = iris.target[mask]
    
    le = LabelEncoder()
    le.fit(y)
    y = le.transform(y)
    
    sklearn_cls = GradientBoostingClassifier(n_estimators=n_estimators, max_depth=max_depth, max_leaf_nodes=n_leaves,
                                             learning_rate=learning_rate, verbose=2)
    sklearn_cls.fit(X, y)
    
    lgbm_cls = LGBMClassifier(n_estimators=n_estimators, max_depth=max_depth, num_leaves=n_leaves,
                              learning_rate=learning_rate, verbosity=2)
    
    lgbm_cls.fit(X, y)
    
    fig = plt.figure()
    ax = plt.subplot(1, 2, 1)
    ax.set_title("Sklearn Gradient Boosting")
    
    plot_classification(ax, sklearn_cls, X, y)
    
    ax = plt.subplot(1, 2, 2)
    ax.set_title("LightGBM")
    
    plot_classification(ax, lgbm_cls, X, y)
    
    plt.show()

**Comment by guolinke** [comment]
I guess the data is very small . And you can try min_data=1 and min_bin_in_bin=1

**Comment by MatthiasMichael** [comment]
Thank you so much. Setting min_data_in_bin=1 solved my issue.
